### PR TITLE
New relation integration test for MySQL VM 

### PR DESCRIPTION
**Pull request by paulomach**
# Issue

New relations implementation were lacking integration tests.

# Solution

Add a custom test charm that uses data-platform [requires lib](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/database_requires.py), and all surrounding toolchain to make the test run.

Complexity was deliberately implemented in the test charm code.
The test code itself just deploys a mysql cluster (3 units), the test application charm (2 units), relate them and check statuses.
Data insertion into database primary and further data reading from secondary are done in the charm code, triggered by the relation event generation. If all goes well the status are set and asserted by the test code.
All fields (as implemented) from the relation data are validated.
 

# Context

There are no charms yet using our data-platform standard relations, hence the custom charm to accomplish that. 

# Testing

Implemented integration tests

# Release Notes

- `.github/workflows/ci.yaml` - Add test execution in CI
- `pyproject.toml` - Add asyncio default mode
- `tests/integration/application-charm/*{txt,yaml}` - test charm boilerplate
- `tests/integration/application-charm/lib/charms/data_platform_libs/v0/database_requires.py` - dp-libs requires. Open to suggestion on how to add this just for the test - for the review, just skip this.
- `tests/integration/application-charm/src/charm.py` - test charm code
- `tests/integration/application-charm/src/connector.py` - context manager for `mysql.connector` upstream lib
- `tests/integration/test_database.py` - shameless simple test code
- `tox.ini` - target for new test
- `README.md` - updated readme with new relation

### CODE REVIEW

File: tests/integration/test_database.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    app_charm = await ops_test.build_charm("./tests/integration/application-charm/")
+    config = {"cluster-name": CLUSTER_NAME}
+    await ops_test.model.deploy(
+        charm, application_name=DATABASE_APP_NAME, config=config, num_units=3
+    )
+
+    await ops_test.model.deploy(app_charm, application_name=APPLICATION_APP_NAME, num_units=2)
+
+    # Reduce the update_status frequency until the cluster is deployed
+    await ops_test.model.set_config({"update-status-hook-interval": "10s"})
```
Comment (shayancanonical) [inline tests/integration/test_database.py:42]: could we use `ops_test.model.fast_forward` instead?

File: tox.ini [hunk 2]
```diff
@@ -112,3 +112,15 @@ deps =
...
+description = Run integration tests
+deps =
+    juju
+    mysql-connector-python
+    pytest
+    pytest-operator
+    pytest-order
+    -r{toxinidir}/requirements.txt
+commands =
+    pytest -v --tb native --ignore={[vars]tst_path}unit --log-cli-level=INFO -s {posargs} -m "database_tests"
```
Comment (shayancanonical) [inline tox.ini:126]: missing new line

File: tests/integration/test_database.py (hunk 1 above)
Comment (paulomach) [reply tests/integration/test_database.py:42]: Yep, I've tried, but afaik it's not yet present in the latest release of pytest-operator :disappointed:

**Review by shayancanonical — APPROVED** [review]

File: tests/integration/test_database.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    await ops_test.model.set_config({"update-status-hook-interval": "60m"})
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.database_tests
+async def test_relation(ops_test: OpsTest):
+    await ops_test.model.add_relation(APPLICATION_APP_NAME, f"{DATABASE_APP_NAME}:database")
+
+    await ops_test.model.wait_for_idle(apps=APPS, status="active")
```
Comment (MiaAltieri) [inline tests/integration/test_database.py:79]: lol great description in your PR "shameless simple test code" I assume at some point this file will be populated with actual tests correct?

File: tests/integration/test_database.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    )
+    assert len(ops_test.model.applications[APPLICATION_APP_NAME].units) == 2
+
+    await ops_test.model.set_config({"update-status-hook-interval": "60m"})
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.database_tests
+async def test_relation(ops_test: OpsTest):
```
Comment (MiaAltieri) [inline tests/integration/test_database.py:76]: I know you mentioned these were simple tests so this is just a comment for the future. You'll want to add more robust tests for your relations. feel free to grab inspiration from [mine](https://github.com/canonical/mongodb-operator/blob/main/tests/integration/relation_tests/new_relations/test_charm_relations.py)

File: tests/integration/test_database.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    await ops_test.model.add_relation(APPLICATION_APP_NAME, f"{DATABASE_APP_NAME}:database")
+
```
Comment (MiaAltieri) [inline tests/integration/test_database.py:78]: warning wait for idle might go through without waiting for relation hooks, leading to occassional failures and race conditions see [this convo](https://chat.canonical.com/canonical/pl/nh554o9gcfny8pcgu998bx759a)

**Review by MiaAltieri — CHANGES_REQUESTED** [review]

**Review by zmraul — APPROVED** [review]
Looks great! The comments from Mia about relation waiting issues are worth addressing.

I'm wondering if creating an action on the application-charm to use the database and report results would be a good pattern. This way we can split the relation logic from the usage of the relation afterwards.

File: tests/integration/test_database.py (hunk 3 above)
Comment (paulomach) [reply tests/integration/test_database.py:79]: Yep. We have a whole batch of full blown shakedown tests planned for next features.

**Review by marceloneppel — APPROVED** [review]

File: tests/integration/test_database.py (hunk 4 above)
Comment (paulomach) [reply tests/integration/test_database.py]: tks for the reference.

File: tests/integration/test_database.py (hunk 5 above)
Comment (paulomach) [reply tests/integration/test_database.py:89]: fixed

Comment (MiaAltieri) [reply tests/integration/test_database.py:89]: looks great, awesome use of `ops_test.model.block_until`

**Review by MiaAltieri — APPROVED** [review]

**Pull request merged**